components/playlist/browser/playlist_media_file_downloader.cc

```diff
@@ -0,0 +1,374 @@
+      playlist_dir_path_.Append(source_media_files_dir_)
+          .Append(GetFileNameStringFromIndex(index));
+  iter->get()->DownloadToFile(
+      url_loader_factory_.get(),
+      base::BindOnce(&PlaylistMediaFileDownloader::OnMediaFileDownloaded,
+                     base::Unretained(this), std::move(iter), index),
+      file_path);
+}
+
+void PlaylistMediaFileDownloader::OnMediaFileDownloaded(
+    SimpleURLLoaderList::iterator iter,
+    int index,
+    base::FilePath path) {
+  // When cancelled, we don't need to more for current job.
+  if (cancelled_)
+    return;
+
+  url_loaders_.erase(iter);
+
+  if (path.empty()) {
+    // This fail is handled during the generation.
+    // See |has_skipped_source_files| in DoGenerateSingleMediaFileOnIOThread().
+    // |has_skipped_source_files| will be set to true.
+    VLOG(1) << __func__ << ": failed to download media file at " << index;
+  }
+
+  remained_download_files_--;
+
+  // If all source files are downloaded, unify them into one media file.
+  if (IsDownloadFinished())
+    StartSingleMediaFileGeneration();
+}
+
+void PlaylistMediaFileDownloader::RequestCancelCurrentPlaylistGeneration() {
+  cancelled_ = true;
+  url_loaders_.clear();
+}
+
+void PlaylistMediaFileDownloader::StartSingleMediaFileGeneration() {
+  base::PostTaskAndReplyWithResult(
+      io_task_runner(), FROM_HERE,
+      base::BindOnce(&DoGenerateSingleMediaFileOnIOThread, playlist_dir_path_,
+                     source_media_files_dir_, unified_media_file_name_,
+                     media_file_source_files_count_),
+      base::BindOnce(&PlaylistMediaFileDownloader::OnSingleMediaFileGenerated,
+                     weak_factory_.GetWeakPtr()));
+}
+
+void PlaylistMediaFileDownloader::OnSingleMediaFileGenerated(int result) {
+  if (cancelled_) {
+    ResetStatus();
+    return;
+  }
+
+  if (result != -1) {
+    base::FilePath media_file_path =
+        playlist_dir_path_.Append(unified_media_file_name_);
+    const std::string media_file_path_utf8 =
+#if defined(OS_WIN)
+        base::UTF16ToUTF8(media_file_path.value());
+#else
+        media_file_path.value();
+#endif
+    const bool partial_ready = result == 1;
+    current_playlist_.SetStringKey(media_file_path_key_, media_file_path_utf8);
+    current_playlist_.SetBoolKey(kPlaylistPartialReadyKey, partial_ready);
+    NotifySucceed(partial_ready);
+  } else {
+    current_playlist_.SetStringKey(media_file_path_key_, "");
+    current_playlist_.SetBoolKey(kPlaylistPartialReadyKey, false);
+    NotifyFail();
+  }
+}
+
+base::SequencedTaskRunner* PlaylistMediaFileDownloader::io_task_runner() {
+  if (!io_task_runner_) {
+    io_task_runner_ = base::CreateSequencedTaskRunner(
+        {base::ThreadPool(), base::MayBlock(), base::TaskPriority::BEST_EFFORT,
+         base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN});
+  }
+  return io_task_runner_.get();
+}
+
+void PlaylistMediaFileDownloader::ResetStatus() {
+  in_progress_ = false;
```

Comment:
should we clear `remained_download_files_` here? And `media_file_source_files_count_` ?